libraries/bezier-rs/src/subpath/manipulators.rs

```diff
@@ -0,0 +1,114 @@
+use super::*;
+use crate::ComputeType;
+
+impl Subpath {
+	/// Calculate the point on the subpath based on the parametric `t`-value provided.
+	/// Expects `t` to be within the inclusive range `[0, 1]`.
+	pub fn add_manipulator_group(&mut self, t: ComputeType) {
+		match t {
+			ComputeType::Parametric(t) => {
+				assert!((0.0..=1.).contains(&t));
+
+				let number_of_curves = self.len_segments() as f64;
+				let scaled_t = t * number_of_curves;
+
+				let target_curve_index = scaled_t.floor() as i32;
+				let target_curve_t = scaled_t % 1.;
+
+				if target_curve_t == 0. || target_curve_t == 1. {
+					return;
+				}
+
+				// The only case where `curve` would be `None` is if the provided argument was 1
+				// But the above if case would catch that, since `target_curve_t` would be 0.
+				let curve = self.iter().nth(target_curve_index as usize).unwrap();
+
+				let [first, second] = curve.split(target_curve_t);
+				let new_group = ManipulatorGroup {
+					anchor: first.end(),
+					in_handle: first.handle_end(),
+					out_handle: second.handle_start(),
+				};
+				let number_of_groups = self.manipulator_groups.len() + 1;
+				self.manipulator_groups.insert((target_curve_index as usize) + 1, new_group);
+				self.manipulator_groups[(target_curve_index as usize) % number_of_groups].out_handle = first.handle_start();
+				self.manipulator_groups[((target_curve_index as usize) + 2) % number_of_groups].in_handle = second.handle_end();
+			}
+			// TODO: change this implementation to Euclidean compute
+			ComputeType::Euclidean(_t) => {}
+			ComputeType::EuclideanWithinError { t: _, epsilon: _ } => todo!(),
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use glam::DVec2;
+
+	#[test]
+	fn add_manipulator_group() {
```

Comment:
Could you break down these tests further and give them more descriptive names based on the conditions they are verifying?

Reply:
I added comments in between each section as descriptions. Is that alright or do you want them to be individual functions?

Reply:
I usually leave a single test case per test because its more informative. That way if an early test fails, you still know the results of the later tests. As a compromise could we at least separate out the open / closed test cases? Also there's some grammatical issues in a few of the comments.